File: Algorithms/p103_Binary_Tree_Zigzag_Level_Order_Traversal/p103_Binary_Tree_Zigzag_Level_Order_Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        res = []
        q = collections.deque([root])
        forward = True
        while q:
            res.append([])
            size = len(q)
            for i in range(size):
                if forward:
                    p = q.popleft()
                    if p.left:
                        q.append(p.left)
                    if p.right:
                        q.append(p.right)
                else:
                    p = q.pop()
                    if p.right:
                        q.appendleft(p.right)
                    if p.left:
                        q.appendleft(p.left)
                res[-1].append(p.val)
            forward = not forward
        return res

File: Algorithms/p103_Binary_Tree_Zigzag_Level_Order_Traversal/p103_Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root) return {};
        vector<vector<int>> res;
        deque<TreeNode*> q;
        bool forward = 1;
        q.push_back(root);
        while(!q.empty()){
            res.push_back({});
            int size = q.size();
            for(int i=0;i<size;++i){
                TreeNode* p;
                if(forward){
                    p = q.front();
                    q.pop_front();
                    if(p->left) q.push_back(p->left);
                    if(p->right) q.push_back(p->right);
                }else{
                    p = q.back();
                    q.pop_back();
                    if(p->right) q.push_front(p->right);
                    if(p->left) q.push_front(p->left);
                }
                res.back().push_back(p->val);
            }
            forward = !forward;
        }
        return res;
    }
};
